webviz_config/webviz_factory_registry.py

```diff
@@ -0,0 +1,84 @@
+from typing import Dict, TypeVar, Type, Optional, Any
+
+from .webviz_instance_info import WebvizInstanceInfo
+
+# pylint: disable=invalid-name
+T = TypeVar("T")
+
+
+class WebvizFactoryRegistry:
+    """Global registry for factories that allows the actual factory instances
+    to be shared between plugins. Also facilitates storage of optional factory
+    settings that are read from the YAML config file for later consumption when
+    an actual factory gets created and added to the registry.
+    The registry is exposed globally through WEBVIZ_FACTORY_REGISTRY below.
+    This is also the reason for the two-stage initialization approach. Note that
+    the registry instance is useless until the initialize() method has been called.
+    """
+
+    def __init__(self) -> None:
+        self._is_initialized: bool = False
+        self._app_instance_info: Optional[WebvizInstanceInfo] = None
+        self._factory_settings_dict: Dict[str, Any] = {}
+        self._factories: Dict[Type, Any] = {}
+
+    def initialize(
+        self,
+        app_instance_info: WebvizInstanceInfo,
+        factory_settings_dict: Optional[Dict[str, Any]],
+    ) -> None:
+        """Does the actual initialization of the object instance.
+        This function will be called as part of the webviz_app.py / jinja2 template.
+        """
+        if self._is_initialized:
+            raise ValueError("Registry already initialized")
+
+        if not isinstance(app_instance_info, WebvizInstanceInfo):
+            raise TypeError("app_instance_info must be of type WebvizInstanceInfo")
+        self._app_instance_info = app_instance_info
+
+        if factory_settings_dict:
+            if not isinstance(factory_settings_dict, dict):
+                raise TypeError("factory_settings_dict must be of type dict")
+            self._factory_settings_dict = factory_settings_dict
+
+        self._is_initialized = True
+
+    def set_factory(self, factory_class: Type[T], factory_obj: T) -> None:
```

Comment:
Have we considered getting the factory class from `.__class__`?

```python
class FactoryClass:
    pass

factory_obj = FactoryClass()

factory_obj.__class__ == FactoryClass  # == true
isinstance(factory_obj, FactoryClass)  # == true
```

Reply:
As far as I understand that would not work here. The intention is that `factory_class` should serve as the lookup key in the registry and that the `factory object` should be an instance of `factory_class` or a subclass of `factory_class`. 
Wouldn't comparing against `.__class__` require an exact match?

Reply:
Ah, ok. Yes, `.__class__` would require an exact match.

From the (to be added) base class, would a given chain of subclasses have _one_ class in the chain to use as key?

Maybe `__bases__` or `__mro__` can be used? I'm probably of more review help after the `CONTRIBUTING.md` section with intended high-level usage in plugin projects. :smile: 

```python
class FactoryClass:
    pass

class SubFactoryClass(FactoryClass):
    pass

sub_factory_obj = SubFactoryClass()

print(sub_factory_obj.__class__.__mro__)  # (<class '__main__.SubFactoryClass'>, <class '__main__.FactoryClass'>, <class 'object'>)
print(sub_factory_obj.__class__.__bases__)  # (<class '__main__.FactoryClass'>,)
```